### PR TITLE
more camera output options

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

- fix some post-processing issues
- provide a means to manage advanced camera composition 

## Solution

- added a `CameraOutputMode` to the `Camera` struct. defines whether the camera retains it's results or flushes them to the output. `CameraOutputMode::Default` will retain results unless the camera is last, in which case it will flush them.
- changed upscaling to write back to the next pass's input texture when output mode is `Retain`, which fixes some issues with post-processing steps being dropped
- calculated the camera's `CameraChainPosition` in `extract_cameras`, which describes whether a camera is first, first after a flush or neither of those, and
- changed ClearColorConfig::Default to do the most appropriate thing in the current camera position
    - first -> clear with the global clear color  
    - first after a flush -> clear with transparent
    - not first -> don't clear
- (fixed bloom alpha as well - this was necessary to make alpha-blending of bloomed output work correctly)

fix https://github.com/bevyengine/bevy/issues/7435
fix https://github.com/bevyengine/bevy/issues/7361 (you would need to add a Flush to the first camera, so that bloom on the second camera doesn't affect the first camera's results)  _EDIT: and remove the `ClearColorConfig::None` from the second camera so it defaults to transparent clear_
fix https://github.com/bevyengine/bevy/issues/5721

note - if you use multiple overlapping cameras in Retain mode, you should disable FXAA and tonemapping (and any other fullscreen effects) on all but the last one.

todo:

 - [ ] use remote reflection for the blend state when #6042 lands
 - [ ] maybe move the camera chain calcs out of extract? they are not very expensive as there are generally not a lot of cameras, but could do this if needed

---

## Migration Guide

this should maintain all existing behaviour - i think if it worked before it'll still work now. On secondary cameras, adding 'ClearColorConfig::None` is no longer required but is not harmful.

### CODE REVIEW

File: crates/bevy_core_pipeline/Cargo.toml [hunk 1]
```diff
@@ -31,3 +31,4 @@ bevy_utils = { path = "../bevy_utils", version = "0.9.0" }
 serde = { version = "1", features = ["derive"] }
 bitflags = "1.2"
 radsort = "0.1"
+wgpu = "0.15.0"
```
Comment (superdump) [inline crates/bevy_core_pipeline/Cargo.toml:34]: newline

File: crates/bevy_render/src/camera/camera.rs [hunk 2]
```diff
@@ -75,6 +75,30 @@ pub struct ComputedCameraValues {
...
+    /// If multiple flushing cameras are used, later cameras may prefer to use `Flush(Some(BlendState::ALPHA_BLENDING))`
+    /// to combine render results with previously flushed results.
+    /// if a camera uses `Flush`, the next camera on the the same target should clear the internal render results with
+    /// `ClearColorConfig::Custom(Color::NONE)` to avoid double writing of the first results on the next flush.
+    Flush(Option<wgpu::BlendState>),
+    /// Keep the current render results internally. A later camera must output with `Flush` for any results to be output.
+    /// This is useful for combining the results of several cameras before applying post-processing to the full set of results.
+    /// If a camera uses `Retain`, the next camera on the same target should make sure NOT to clear the render results by
+    /// using `ClearColorConfig::None`, else they will be discarded.
+    Retain,
```
Comment (superdump) [inline crates/bevy_render/src/camera/camera.rs:101]: I feel like the names don't make obvious what the modes are. As I was reading various parts of the PR I wasn't sure if Flush was going to mean something like discarding or writing out what is present. Similarly Retain feels like it could mean keep the results (as in write them out) when compared to Flush, if Flush were understood to mean discard.

How about renaming `Flush` to `Blend` or perhaps `Combine` (I prefer `Blend`)?

File: crates/bevy_core_pipeline/src/bloom/mod.rs [hunk 3]
```diff
@@ -512,7 +512,11 @@ impl FromWorld for BloomPipelines {
                                 dst_factor: BlendFactor::One,
                                 operation: BlendOperation::Add,
                             },
-                            alpha: BlendComponent::REPLACE,
+                            alpha: BlendComponent {
+                                src_factor: BlendFactor::One,
+                                dst_factor: BlendFactor::One,
+                                operation: BlendOperation::Max,
+                            },
```
Comment (superdump) [inline crates/bevy_core_pipeline/src/bloom/mod.rs:519]: For what was this needed?

**Comment by superdump** [comment]
Adding this to 0.10 as it looks to clear up quite a few camera composition and post-processing issues.

File: crates/bevy_render/src/camera/camera.rs (hunk 2 above)
Comment (robtfm) [reply crates/bevy_render/src/camera/camera.rs:101]: hmm... i was trying to draw an analogy with stream flushing (like stdout). my problem with combine and blend is that they don't really make you feel like you *need* to do it to get results. but its true that flush can also mean "clear" so it isn't great.

maybe `Finish` and `Continue`..?

File: crates/bevy_core_pipeline/src/bloom/mod.rs (hunk 3 above)
Comment (robtfm) [reply crates/bevy_core_pipeline/src/bloom/mod.rs:519]: bloom currently outputs alpha of the bloom layer (which is typically small) and destroys the alpha of the input. this made blending behave badly in multi-flush scenarios with bloom.

File: crates/bevy_render/src/camera/camera.rs (hunk 2 above)
Comment (kurtkuehnert) [reply crates/bevy_render/src/camera/camera.rs:101]: Yeah, the naming isn't intuitive.

`Finish` and `Continue` are clearer, but continue still seems relatively unspecific. 

Maybe the combination of `Finish` and `Retain` (or even `Gather`?) might be an option.

File: crates/bevy_core_pipeline/src/clear_color.rs [hunk 4]
```diff
@@ -1,18 +1,42 @@
...
     Default,
     Custom(Color),
     None,
 }
 
+impl ClearColorConfig {
+    pub fn load_op<V: From<bevy_render::color::Color>>(
+        &self,
+        chain_position: CameraChainPosition,
+        default_color: &Color,
```
[the remark this answers is not in the document]
Comment (robtfm) [reply crates/bevy_core_pipeline/src/clear_color.rs]: done

**Review by kurtkuehnert — APPROVED** [review]
I am not super familiar with the camera and postprocessing code, but the code quality is great and the feature is easy enough to understand from my point of view. LGTM.

**Comment by cart** [comment]
This feels a bit weird to me for a number of reasons. Took me awhile to sort out _why_ it felt wrong, but I think I finally consolidated my thoughts: 

1. It introduces the concept of a "base" texture (aka the "first" texture). This makes the mental model more complicated. I think all that should matter is "which texture is the current texture" and "which texture is the next texture". Within the context of a single camera, this is all that matters. Across cameras, which one is the "current texture" should be an implementation detail. Writing back to a "canonical base intermediate texture" should never be a thing. It seems like this is attempting to solve the "we implicitly reset the RenderTarget's main texture to the first texture between cameras, because each camera has its own index" problem. It seems like the real solution here is to share that index across cameras. That is a "render target property" not a "camera ViewTarget property". Additionally, overloading the "upscaling node" to do this "writeback to base" also feels wrong / unclear. But that problem is solved if we just stop doing writeback.
2.  The new default "clear color is dependent on CameraChainPosition" behavior is complicated / implicit / hard to follow. Ditto for the default CameraOutputMode.

This also doesn't seem to fix #7361. I think I did the recommended config correctly?
(`Finish(None)` on the first camera, `ClearColorConfig::None` on the second)
https://gist.github.com/cart/4bdae302232d67ad53cdb3293875c5c2. This feels like a bloom-specific implementation issue? It seems like the "cameras are a direct path from logical views to pixels in RenderTargets" mental model (the one on main) on its own should "sandbox" cameras, provided the second bloom camera does a clear in the main pass.

I _think_ this is my preferred solution (starting from the current state on `main`):
1. Make the "current active intermediate texture index" a shared RenderTarget property (across cameras with the same render target). An `Arc<AtomicUsize>` assigned in `prepare_view_targets` should do the trick.
2. Add a simplified CameraOutputMode that just has `CameraOutputMode::Skip` (skips the upscaling pass entirely) and `CameraOutputMode::Write(Option<BlendMode>)`, which writes the output in the upscaling pass using the given blend mode. This should probably default to `CameraOutputMode::Write(Some(BlendMode::ALPHA_BLENDING))`, but we can discuss that.
3. Leave ClearColor untouched. No implicit default stuff.

This way, whenever you access the `main_texture` for a RenderTarget (at any point in the pipeline across cameras), it will always be valid. If your first camera draws something and your second camera doesn't clear, the second camera will have access to that state, enabling "aggregate camera output" post processing. This removes the need for weird cross-camera statefulness, "automatic implicit defaults", or unnecessary texture writebacks to make things work as expected.

For example:
```
// draw the scene, skip upscaling, default clear color
CameraX: order = 0, CameraOutputMode::Skip, clearcolor::default
// draw another part of the scene, skip upscaling, retain outputs from X because we dont clear,
CameraY: order = 1, CameraOutputMode::Skip, clearcolor::None 
// draw nothing, but apply post processing effects to retained outputs from X and Y because we dont clear,
// writes to actual render target because we use the default CameraOutputMode::Write mode 
CameraZ: order = 2, CameraOutputMode::default(), clearcolor:None
```

**Comment by robtfm** [comment]
> 1. "which texture is the current texture" and "which texture is the next texture". Within the context of a single camera, this is all that matters

this is made trickier with msaa. after post-processing we are on a non-msaa texture and the next pass wants to start from the multisampled texture so a copy-back is really required. i thought doing the simple thing and always copying back was the easiest to follow (and this will need reworking if/when we do msaa-per-camera in future anyway).

> This also doesn't seem to fix #7361. I think I did the recommended config correctly?

in this case you don't want to set `ClearColorConfig::None` as that means "don't clear" so bloom still affects the first shape. without it it'll default to `Custom(None)` i.e. clear to transparent and will work -- but you'll need to turn up the bloom power as well since the second camera now doesn't have the 0.1 background colour to bounce the bloom off.

> * Make the "current active intermediate texture index" a shared RenderTarget property (across cameras with the same render target). An `Arc<AtomicUsize>` assigned in `prepare_view_targets` should do the trick.

with msaa i think copy back is required

> * Add a simplified CameraOutputMode that just has `CameraOutputMode::Skip` (skips the upscaling pass entirely) and `CameraOutputMode::Write(Option<BlendMode>)`, which writes the output in the upscaling pass using the given blend mode. This should probably default to `CameraOutputMode::Write(Some(BlendMode::ALPHA_BLENDING))`, but we can discuss that.

i also changed the upscaling to `Load` instead of `Clear` so that multiple passes can be composed, so you need the first flushing cam to be `Finish(None)` or `OVERWRITE` or thereabouts to clear the previous frame data.

> * Leave ClearColorConfig untouched. No implicit default stuff.

happy to remove that behaviour if you prefer. i added it since the correct choice isn't always obvious (maybe we should rename `ClearColorConfig::None` so it's not so easy to mix up with `ClearColorConfig::Custom(None)`), and wanted to make things work as well as possible without extra config.

**Comment by cart** [comment]
> this is made trickier with msaa. after post-processing we are on a non-msaa texture and the next pass wants to start from the multisampled texture so a copy-back is really required. i thought doing the simple thing and always copying back was the easiest to follow (and this will need reworking if/when we do msaa-per-camera in future anyway).

Arg yeah I completely left that out of my mental model. My bad! Definitely complicates things. I'll ponder this for a bit and get back to you.

> in this case you don't want to set ClearColorConfig::None as that means "don't clear" so bloom still affects the first shape. without it it'll default to Custom(None) i.e. clear to transparent and will work -- but you'll need to turn up the bloom power as well since the second camera now doesn't have the 0.1 background colour to bounce the bloom off.

That does indeed work!

> i also changed the upscaling to Load instead of Clear so that multiple passes can be composed, so you need the first flushing cam to be Finish(None) or OVERWRITE or thereabouts to clear the previous frame data.

Hmm that makes sense. I _think_ making this explicit (and i guess defaulting to overwrite to make the normal case work) might be the move. I'll add it to the list of constraints that I missed and get back to you :)

**Comment by cart** [comment]
My non-serious suggestion is that we remove MSAA in the interest of pipeline purity :)

**Comment by cart** [comment]
Ok so I've been experimenting / poking / prodding / thinking today and it definitely does seem like writeback is required (for MSAA). Here is my "next" proposal:

1. Add "msaa writeback" (actual name TBD) as an explicit "msaa camera feature" and default to `msaa_writeback: true` for each camera. If this is true, a camera has MSAA enabled, and it isn't the first camera for the target, add a writeback _before_ the main pass for that camera. Note that this moves writeback from the "previous" camera to the "camera that needs MSAA textures to contain the last camera's results". This makes this behavior explicit and scopes it specifically to MSAA.
2.  Use this CameraOutputMode, which makes write behaviors explicit and directly configurable.
```rust
#[derive(Debug, Clone, Copy)]
pub enum CameraOutputMode {
    /// Writes the camera output to configured render target.
    Write {
        /// The blend state that will be used by the pipeline that writes the intermediate render textures to the final render target texture.
        blend_state: Option<BlendState>,
        /// The color attachment load operation that will be used by the pipeline that writes the intermediate render textures to the final render
        /// target texture.
        color_attachment_load_op: wgpu::LoadOp<wgpu::Color>,
    },
    /// Skips writing the camera output to the configured render target. The output will remain in the
    /// Render Target's "intermediate" textures, which a camera with a higher order should write to the render target
    /// using [`CameraOutputMode::Write`]. The "skip" mode can easily prevent render results from being displayed, or cause
    /// them to be lost. Only use this if you know what you are doing!
    Skip,
}
```

**Comment by cart** [comment]
Also this is just a proposal. If I've (once again) missed something important please let me know. Very open to feedback!

**Comment by cart** [comment]
Closing in favor of #7671
